checkout: Expose new place order process via REST

danielpoe opened the pull request:
Add rest apis to be able to place the current cart

see http://editor.swagger.io/?url=https://raw.githubusercontent.com/i-love-flamingo/flamingo-commerce/more-rest-apis/docs/openapi/swagger.json

carstendietrich commented:
Looks promising, will have an in detail look later today.

Guess we would need some integration tests for sure.

Also we are lacking rest routes for the following GraphQL mutations:
```graphql
# Gets the last stored place order state and ensures that the state machine proceeds, non blocking
Commerce_Checkout_RefreshPlaceOrder: Commerce_Checkout_PlaceOrderContext!
# Gets the most recent place order state by waiting for the state machine to proceed, therefore blocking
Commerce_Checkout_RefreshPlaceOrderBlocking: Commerce_Checkout_PlaceOrderContext!
```

Without those we can't proceed in the state machine. And we need definitely add them.

danielpoe commented:
> Looks promising, will have an in detail look later today.
> 
> Guess we would need some integration tests for sure.
> 
> Also we are lacking rest routes for the following GraphQL mutations:
> 
> ```graphql
> # Gets the last stored place order state and ensures that the state machine proceeds, non blocking
> Commerce_Checkout_RefreshPlaceOrder: Commerce_Checkout_PlaceOrderContext!
> # Gets the most recent place order state by waiting for the state machine to proceed, therefore blocking
> Commerce_Checkout_RefreshPlaceOrderBlocking: Commerce_Checkout_PlaceOrderContext!
> ```
> 
> Without those we can't proceed in the state machine. And we need definitely add them.

Good point - I added the API
I also added integration test :-)

carstendietrich commented on checkout/module.go:
```diff
@@ -190,3 +191,27 @@ func (m *Module) Depends() []dingo.Module {
+	registry.Route("/api/v1/checkout/placeorder/cancel", "checkout.api.placeorder.cancel")
+	registry.HandleGet("checkout.api.placeorder.cancel", r.apiController.CancelPlaceOrderAction)
+
+	registry.Route("/api/v1/checkout/placeorder/refreshplaceorder", "checkout.api.placeorder.refreshplaceorder")
```
why duplication of placeorder?
```suggestion
	registry.Route("/api/v1/checkout/placeorder/refresh", "checkout.api.placeorder.refresh")
```

carstendietrich commented on checkout/module.go:
```diff
@@ -190,3 +191,27 @@ func (m *Module) Depends() []dingo.Module {
+	registry.Route("/api/v1/checkout/placeorder/refreshplaceorder", "checkout.api.placeorder.refreshplaceorder")
+	registry.HandlePost("checkout.api.placeorder.refreshplaceorder", r.apiController.RefreshPlaceOrderAction)
+
+	registry.Route("/api/v1/checkout/placeorder/refreshplaceorderblocking", "checkout.api.placeorder.refreshplaceorderblocking")
```
maybe:
```suggestion
	registry.Route("/api/v1/checkout/placeorder/refresh/blocking", "checkout.api.placeorder.refresh.blocking")
```

carstendietrich commented on test/integrationtest/projecttest/tests/restapi/placeorderszenario_test.go:
```diff
@@ -0,0 +1,44 @@
+// +build integration
```
please rename file to just `placeorder_test.go`

carstendietrich commented on test/integrationtest/projecttest/tests/restapi/placeorderszenario_test.go:
```diff
@@ -0,0 +1,44 @@
+	response = e.GET("/api/v1/checkout/placeorder").Expect()
+	response.Status(200).JSON().Object().Value("FailedReason").String().Equal("")
+	response.Status(200).JSON().Object().Value("State").String().Equal("Success")
+
```
I guess we should at least call each route in this test. Currenty GET /placeorder/ and POST /placeorder/refreshing is missing.

carstendietrich commented:
> Looks good so far, fixed some typos my self and added missing Swagger summaries.
> Route naming and test coverage could be improved before merging this.

Besides that module readme needs to be updated, previously new process was GraphQL exclusive. CHANGELOG.md in project root should also contain this new additions

carstendietrich commented on checkout/interfaces/controller/apicontroller.go:
```diff
@@ -0,0 +1,244 @@
+		Cart         *cart.Cart
+		OrderInfos   *placedOrderInfos
+		State        string
+		StateData    interface{}
```
maybe using type StateData here as well, is also type interface{} but would match the core place order context

carstendietrich commented on checkout/interfaces/controller/apicontroller.go:
```diff
@@ -0,0 +1,244 @@
+// @Router /api/v1/checkout/placeorder [delete]
+func (c *APIController) ClearPlaceOrderAction(ctx context.Context, r *web.Request) web.Result {
+	err := c.placeorderHandler.ClearPlaceOrder(ctx)
+	if err != nil {
```
should we use another status code whenever there is no place order process running? in this case we currently return `{"Code":"500","Message":"ErrNoPlaceOrderProcess"}`
